### PR TITLE
Added medication administration specs, model and viewset

### PR DESCRIPTION
**Pull request by khavinshankar**
- Added medication administration specs, model and viewset
- Follows specs in https://ohcnetwork.notion.site/Medicine-Administration-12ae6e667ece80f3a6fbe12af2ef729a, removed site and route from the root as they were also repeated in dosage

### CODE REVIEW

File: care/emr/api/viewsets/medication_administration.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    def filter_occurrence_period_end(self, queryset, name, value):
+        """
+        Filters by occurrence_period.end using DateTimeFromToRangeFilter.
+        """
+
+        if value:
+            queryset = (
+                queryset.annotate(
+                    occurrence_end_datetime=Cast(
```
Comment (vigneshhari) [inline care/emr/api/viewsets/medication_administration.py]: Nope, If you need filters, they need to be moved out of the json, no point in creating complicated queries.

File: care/emr/api/viewsets/medication_administration.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        return queryset
+
+
+class MedicationAdministrationViewSet(EMRModelViewSet):
+    database_model = MedicationAdministration
+    pydantic_model = MedicationAdministrationSpec
+    pydantic_read_model = MedicationAdministrationReadSpec
+    filterset_class = MedicationAdministrationFilter
+    filter_backends = [filters.DjangoFilterBackend]
+    CREATE_QUESTIONNAIRE_RESPONSE = False
```
Comment (vigneshhari) [inline care/emr/api/viewsets/medication_administration.py]: Questionnaire response needs to be created, this needs to be added in updates.

File: care/emr/resources/medication/administration/spec.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    entered_in_error = "entered_in_error"
+    stopped = "stopped"
+    in_progress = "in_progress"
+    on_hold = "on_hold"
+    unknown = "unknown"
+    cancelled = "cancelled"
+
+
+# https://build.fhir.org/valueset-reason-medication-not-given-codes.html#4.4.1.194
+class MedicationAdministrationStatusReason(str, Enum):
```
Comment (vigneshhari) [inline care/emr/resources/medication/administration/spec.py]: This is a valueset that needs to be fetched from snowstorm, this cannot be stored as a valueset unless there is business logic associated to it in code.

File: care/emr/resources/medication/administration/spec.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+class MedicationAdministrationPerformerFunction(str, Enum):
+    performer = "performer"
+    verifier = "verifier"
+    witness = "witness"
+
+
+class MedicationAdministrationPerformer(BaseModel):
+    actor: UUID4 = Field(
+        ...,
```
Comment (vigneshhari) [inline care/emr/resources/medication/administration/spec.py]: ... is not needed.

File: care/emr/resources/medication/administration/spec.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+
+class MedicationAdministrationPerformerFunction(str, Enum):
+    performer = "performer"
+    verifier = "verifier"
+    witness = "witness"
+
+
+class MedicationAdministrationPerformer(BaseModel):
+    actor: UUID4 = Field(
```
Comment (vigneshhari) [inline care/emr/resources/medication/administration/spec.py:47]: I am highly skeptical of storing a UUID without any relation to what the UUID is and not validating it.

**Review by vigneshhari — CHANGES_REQUESTED** [review]

File: care/emr/resources/medication/administration/spec.py (hunk 5 above)
Comment (khavinshankar) [reply care/emr/resources/medication/administration/spec.py:47]: It always maps to User, added validation but it is still stored as JSON

File: care/emr/resources/medication/administration/spec.py (hunk 4 above)
Comment (khavinshankar) [reply care/emr/resources/medication/administration/spec.py]: When there is a performer, actor is a required field

Comment (vigneshhari) [reply care/emr/resources/medication/administration/spec.py]: The three dots are not needed

File: care/emr/resources/medication/administration/spec.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    def perform_extra_deserialization(self, is_update, obj):
+        if not is_update:
+            obj.encounter = PatientConsultation.objects.get(
+                external_id=self.encounter
+            )  # Needs more validation
+            obj.patient = obj.encounter.patient
+            obj.request = MedicationRequest.objects.get(external_id=self.request)
+
+
+class MedicationAdministrationReadSpec(BaseMedicationAdministrationSpec):
```
Comment (vigneshhari) [inline care/emr/resources/medication/administration/spec.py:225]: List does not need created_by and updated_by, move those to retrieve spec

File: care/emr/resources/medication/request/spec.py [hunk 7]
```diff
@@ -33,6 +33,10 @@ class MedicationRequestStatus(str, Enum):
     draft = "draft"
     unknown = "unknown"
 
+    @classmethod
```
Comment (vigneshhari) [inline care/emr/resources/medication/request/spec.py]: Remove this

File: care/emr/api/viewsets/medication_request.py [hunk 8]
```diff
@@ -8,15 +13,26 @@
...
     MedicationRequestSpec,
+    MedicationRequestStatus,
 )
 from care.emr.resources.questionnaire.spec import SubjectType
 from care.security.authorization import AuthorizationController
 
 
 class MedicationRequestFilter(filters.FilterSet):
     encounter = filters.UUIDFilter(field_name="encounter__external_id")
+    status = filters.MultipleChoiceFilter(
```
Comment (vigneshhari) [inline care/emr/api/viewsets/medication_request.py]: no need for choices, just do an iexact match.

File: care/emr/api/viewsets/medication_request.py [hunk 9]
```diff
@@ -44,5 +60,26 @@ def get_queryset(self):
...
+        responses={200: MedicationRequestReadSpec},
+        tags=["medication_request"],
+    )
+    @action(detail=True, methods=["POST"])
+    def discontinue(self, request, *args, **kwargs):
+        data = MedicationRequestDiscontinueRequest(**request.data)
+        request: MedicationRequest = self.get_object()
+
+        request.status = MedicationRequestStatus.ended
+        request.status_changed = datetime.now(UTC)
```
Comment (vigneshhari) [inline care/emr/api/viewsets/medication_request.py]: use timezone.now()
from django.utils import timezone

File: care/emr/api/viewsets/medication_request.py [hunk 10]
```diff
@@ -44,5 +60,26 @@ def get_queryset(self):
...
+        request.status_reason = data.status_reason
+        request.save()
+
+        return Response(
+            self.get_read_pydantic_model()
+            .serialize(request)
+            .model_dump(exclude=["meta"]),
```
Comment (vigneshhari) [inline care/emr/api/viewsets/medication_request.py]: use to_json()

File: care/emr/api/viewsets/medication_request.py [hunk 11]
```diff
@@ -8,15 +13,26 @@
...
 
 
 class MedicationRequestFilter(filters.FilterSet):
     encounter = filters.UUIDFilter(field_name="encounter__external_id")
+    status = filters.MultipleChoiceFilter(
+        field_name="status", choices=MedicationRequestStatus.choices()
+    )
+    is_prn = filters.BooleanFilter(method="filter_as_needed_boolean")
+
+    def filter_as_needed_boolean(self, queryset, name, value):
```
Comment (vigneshhari) [inline care/emr/api/viewsets/medication_request.py]: We will not query JSON, rebuild the model or remove the filter.

**Review by vigneshhari — CHANGES_REQUESTED** [review]

File: care/emr/api/viewsets/medication_request.py (hunk 11 above)
Comment (khavinshankar) [reply care/emr/api/viewsets/medication_request.py]: Created a new field in the model to record as_needed_boolean, which was initially captured in dosage instructions, a json field

File: care/emr/resources/medication/administration/spec.py (hunk 6 above)
Comment (khavinshankar) [reply care/emr/resources/medication/administration/spec.py:225]: We are using created_by in the list spec in the frontend,

image


_Actually this is similar to other specs, most of the list specs have them._

Comment (vigneshhari) [reply care/emr/resources/medication/administration/spec.py:225]: We will clean up all specs with created_by in them, the joins does not make sense. If updated by is not being used remove it.

**Review by vigneshhari — APPROVED** [review]

**Pull request merged**